### ISSUE TITLE
Clean up SOF FW booting in driver

### ISSUE THREAD

**Issue by keyonjie** [thesofproject/linux#1179]
In general, we have several flags which are FW booting related in struct snd_sof_dev, that is .boot_complete, .boot_wait, .code_loading, .waitq, we need to clean up them, e.g. to use only one or two, we need define clear steps and states during FW booting(until the FW_READY message received and boot finished).

1. try if we can abandon the use of sdev->waitq and sdev->code_loading. please make sure it can works on cAVS 1.5/1.8/2.0+ platforms.
The thought is from: https://github.com/thesofproject/linux/pull/916#issuecomment-490823406

2. change boot_complete to be boot_state, and it can be enumerated as BOOT_INIT, BOOT_ONGOING, BOOT_FAILED, BOOT_SUCCESS.
And then in the place boot failed, set boot_state to BOOT_FAILED, in IPC handling, handle fw_ready message only at BOOT_ONGOING stage.
The thought here is from: https://github.com/thesofproject/linux/pull/1174#issuecomment-528167520

**Comment by plbossart** [comment]
@keyonjie I think we implemented everything listed in this enhancement, can we close?

**Comment by aiChaoSONG** [comment]
@plbossart  We still need to remove some redundant code:
- hda_dsp_ipc_is_sof in hda-ipc.c can be removed, because FW_PURGE will never propagate to IRQ thread, so this function for filter it out can be safely removed.
- wait_event_timeout(sdev->waitq, !sdev->code_loading,  HDA_DSP_CL_TRIGGER_TIMEOUT); the wait on code_loading in hda_loader.c line182 always return 0, which indicate timeout. so sdev->waitq and sdev->code_loading can be removed, too.

**Comment by keyonjie** [comment]
@plbossart I think the item 2 had already done by @ranj063 , for item 1, @aiChaoSONG can you re-activate and proceed your PR please?

**Comment by aiChaoSONG** [comment]
@keyonjie The PR for item 1 is here #1854.

**Comment by keyonjie** [comment]
@aiChaoSONG good, can you do the item 2 you listed above also?

**Comment by aiChaoSONG** [comment]
@keyonjie PR for item 2 is here: https://github.com/thesofproject/linux/pull/1875

**Comment by aiChaoSONG** [comment]
closed for all done in the list
